core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -151,16 +151,22 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     const Kokkos::Experimental::SYCL& space = policy.space();
     Kokkos::Experimental::Impl::SYCLInternal& instance =
         *space.impl_internal_space_instance();
-    sycl::queue& q = *instance.m_queue;
-
-    auto result_ptr = static_cast<pointer_type>(
-        sycl::malloc(sizeof(*m_result_ptr), q, sycl::usm::alloc::shared));
+    using ReductionResultMem =
+        Experimental::Impl::SYCLInternal::ReductionResultMem;
+    ReductionResultMem& reductionResultMem = instance.m_reductionResultMem;
```

Comment:
In hindsight and wrt to #3671, I am not quite sure if we want to treat the result ptr the same as the kernels. In particular, I am not sure if we really want to store the result in `shared` rather than `device` space but I am fine with addressing that later.